### PR TITLE
feat(app-platform): Issue Link UI

### PR DESCRIPTION
**Pull request by mnoble**
**TL;DR:** Adds the necessary UI components (and some backend pieces) to render a Sentry App's "Issue Link" integration alongside the explicit ones for JIRA/GitHub/Azure/etc.

## Summary

This essentially adds copies of the existing "External Issue" components. It adds copies, instead of modifying the existing ones, because the existing ones are so tightly coupled with Integrations that it'd be a nightmare to tease apart, but also very messy (I tried).

## Preparing Sentry App Components

When we request the list of Sentry App Components for an Org, we preemptively "prepare" them. Many of these components require making a request to the Sentry App service to get things like `SelectField` options. We do this ahead of time, instead of JIT, so that if something goes wrong in the process, we can gracefully handle the error by simply not rendering the React component at all.

## Not Using `getEndpoints()`

To retrieve the above, prepared, data we explicitly make a request using `this.api` inside of `ExternalIssuesList`, instead of using `getEndpoints`. We do this so that we can control what happens when an error arises (since this process potentially includes two distributed systems). When this does happen, we simply don't render anything.

Handling every error condition when there's two systems, one of which we have no control over, is not realistically achievable, so we bank on this conservative approach.

## Minor Changes

- Moved `ExternalIssueForm` into it's own file.
- Renamed `IssueSyncElement` to `IssueSyncListElement` to match its filename and how other components were consuming it.
- Added the ability to `filter` Sentry App Components via the API.

### CODE REVIEW

File: src/sentry/api/serializers/models/sentry_app_component.py [hunk 1]
```diff
@@ -13,5 +13,8 @@ def serialize(self, obj, attrs, user):
             'uuid': six.binary_type(obj.uuid),
             'type': obj.type,
             'schema': obj.schema,
-            'sentryAppId': obj.sentry_app_id,
```
Comment (markstory) [inline src/sentry/api/serializers/models/sentry_app_component.py]: Do folks ever need to go from a component response back to the app? Having the ID lets you cook up URLs back to the sentry app.

File: src/sentry/mediators/external_requests/select_requester.py [hunk 2]
```diff
@@ -34,7 +34,7 @@ def call(self):
 
     def _build_url(self):
         domain = urlparse(self.sentry_app.webhook_url).netloc
-        url = u'https://{}{}'.format(domain, self.uri)
+        url = u'http://{}{}'.format(domain, self.uri)
```
Comment (markstory) [inline src/sentry/mediators/external_requests/select_requester.py]: Why the change? Should you be getting the protocol from the webhook URL instead?

File: src/sentry/static/sentry/app/components/group/externalIssueActions.jsx [hunk 3]
```diff
@@ -325,6 +153,104 @@ class ExternalIssueActions extends AsyncComponent {
...
+    if (!this.props.externalIssue) {
+      this.setState({showModal: true});
+    }
+  }
+
+  get link() {
+    const {sentryAppComponent, externalIssue} = this.props;
+
+    let url = '#';
+    let displayName = `Link ${sentryAppComponent.sentryApp.name} Issue`;
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/group/externalIssueActions.jsx]: Should you be using tct() here instead?

File: src/sentry/static/sentry/app/components/group/externalIssueActions.jsx [hunk 4]
```diff
@@ -325,6 +153,104 @@ class ExternalIssueActions extends AsyncComponent {
...
+
+    if (externalIssue) {
+      url = externalIssue.url;
+      displayName = externalIssue.displayName;
+    }
+
+    return (
+      <React.Fragment>
+        <IntegrationIcon src={`icon-github`} />
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/group/externalIssueActions.jsx]: Is this going to change later?

Comment (mnoble) [reply src/sentry/static/sentry/app/components/group/externalIssueActions.jsx]: Heh, yea, it is. The component hard fails if it can't find the icon, so I just put something in here temporarily.

File: src/sentry/mediators/external_requests/select_requester.py (hunk 2 above)
Comment (mnoble) [reply src/sentry/mediators/external_requests/select_requester.py]: Just to test it out locally. Yea, this should be based on their URL.

File: src/sentry/api/serializers/models/sentry_app_component.py (hunk 1 above)
Comment (mnoble) [reply src/sentry/api/serializers/models/sentry_app_component.py]: Yea, it's probably worth having ID in the embedded object.

File: src/sentry/static/sentry/app/components/group/externalIssueActions.jsx (hunk 4 above)
Comment (MeredithAnya) [reply src/sentry/static/sentry/app/components/group/externalIssueActions.jsx]: You can use `src="icon-generic-box"` instead

File: src/sentry/api/endpoints/sentry_app_components.py [hunk 5]
```diff
@@ -21,13 +24,33 @@ def get(self, request, sentry_app):
 class OrganizationSentryAppComponentsEndpoint(OrganizationEndpoint):
     @requires_feature('organizations:sentry-apps')
     def get(self, request, organization):
+        try:
+            project = Project.objects.get(
+                id=request.GET['projectId'],
+                organization_id=organization.id,
+            )
+        except Project.DoesNotExist:
+            return Response('[]')
```
Comment (MeredithAnya) [inline src/sentry/api/endpoints/sentry_app_components.py]: does this have to be a string? can we do `Response([])`?

File: src/sentry/static/sentry/app/components/group/externalIssueForm.jsx [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    integration: PropTypes.object.isRequired,
+    action: PropTypes.oneOf(['link', 'create']),
+    onSubmitSuccess: PropTypes.func.isRequired,
+  };
+
+  shouldRenderBadRequests = true;
+
+  getEndpoints() {
+    const {integration} = this.props;
+    return [['integrationDetails', integration.url]];
```
Comment (MeredithAnya) [inline src/sentry/static/sentry/app/components/group/externalIssueForm.jsx]: I'm not sure where `integration.url` is coming from but this failed for me so I'd put back
```
return [
      [
        'integrationDetails',
        `/groups/${group.id}/integrations/${integration.id}/?action=${action}`,
      ],
]
```

**Review by MeredithAnya — CHANGES_REQUESTED** [review]

File: src/sentry/static/sentry/app/stores/sentryAppInstallationsStore.jsx [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    return this.items;
+  },
+
+  load(items) {
+    this.items = items;
+    this.trigger(items);
+  },
+
+  get(slug) {
+    return this.items.find(item => item.slug === slug);
```
Comment (MeredithAnya) [inline src/sentry/static/sentry/app/stores/sentryAppInstallationsStore.jsx]: should this be using the `item.uuid` here?

**Review by MeredithAnya — APPROVED** [review]

**Comment by sehr-m** [comment]
On it! We are reviewing the PR and will provide feedback shortly.
<sub>Reference ID: `1`</sub>
